Exception when using fakeredis on Py3

Here is a full stacktrace:

```
[2016-08-22 16:37:39,139][ERROR] django.request 47831-123145312813056 base.py:handle_uncaught_exception:284 Internal Server Error: /en/
Traceback (most recent call last):
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/django/core/handlers/base.py", line 174, in get_response
    response = self.process_exception_by_middleware(e, request)
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/django/core/handlers/base.py", line 172, in get_response
    response = response.render()
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/django/template/response.py", line 160, in render
    self.content = self.rendered_content
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/django/template/response.py", line 137, in rendered_content
    content = template.render(context, self._request)
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/django/template/backends/django.py", line 95, in render
    return self.template.render(context)
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/django/template/base.py", line 206, in render
    return self._render(context)
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/django/test/utils.py", line 92, in instrumented_test_render
    return self.nodelist.render(context)
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/django/template/base.py", line 992, in render
    bit = node.render_annotated(context)
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/django/template/base.py", line 959, in render_annotated
    return self.render(context)
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/django/template/loader_tags.py", line 173, in render
    return compiled_parent._render(context)
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/django/test/utils.py", line 92, in instrumented_test_render
    return self.nodelist.render(context)
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/django/template/base.py", line 992, in render
    bit = node.render_annotated(context)
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/django/template/base.py", line 959, in render_annotated
    return self.render(context)
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/django/template/loader_tags.py", line 69, in render
    result = block.nodelist.render(context)
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/django/template/base.py", line 992, in render
    bit = node.render_annotated(context)
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/django/template/base.py", line 959, in render_annotated
    return self.render(context)
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/django/template/library.py", line 201, in render
    output = self.func(*resolved_args, **resolved_kwargs)
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/cachalot/templatetags/cachalot.py", line 18, in get_last_invalidation
    return get_last_invalidation_function(*tables_or_models, **kwargs)
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/cachalot/api.py", line 107, in get_last_invalidation
    cache_alias, db_alias).get_many(table_cache_keys).values()
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/django_redis/cache.py", line 32, in _decorator
    return method(self, *args, **kwargs)
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/django_redis/cache.py", line 107, in get_many
    return self.client.get_many(*args, **kwargs)
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/django_redis/client/default.py", line 337, in get_many
    results = client.mget(*new_keys)
  File "/Users/si14/.virtualenvs/melsciwebpy3/lib/python3.5/site-packages/fakeredis.py", line 381, in mget
    for key in all_keys:
TypeError: 'CacheKey' object is not iterable
```

After some poking around, I think the reason is that `django-redis` uses `CacheKey` as a key (https://github.com/niwinz/django-redis/blob/master/django_redis/util.py#L33), while `fakeredis` has a lot of checks like `isinstance(keys, string_types)` (e.g.  https://github.com/jamesls/fakeredis/blob/master/fakeredis.py#L1454-L1463) `CacheKey` isn't a subclass of `str`, so `fakeredis` breaks.

--- a/.travis.yml
+++ b/.travis.yml
@@ -15,6 +15,7 @@ install:
   - pip install mock
   - pip install msgpack-python>=0.4.6
   - pip install redis>=2.10.0
+  - pip install fakeredis
 
 script:
   - (cd tests && python runtests.py)

--- a/django_redis/util.py
+++ b/django_redis/util.py
@@ -21,7 +21,7 @@ else:
     integer_types = (int,)
 
 
-class CacheKey(object):
+class CacheKey(str):
     """
     A stub string class that we can use to check if a key was created already.
     """

--- a/tests/redis_backend_testapp/tests.py
+++ b/tests/redis_backend_testapp/tests.py
@@ -28,6 +28,9 @@ from django_redis.serializers import json as json_serializer
 from django_redis.serializers import msgpack as msgpack_serializer
 
 
+FAKE_REDIS = settings.CACHES["default"]["OPTIONS"].get("REDIS_CLIENT_CLASS") \
+             == "fakeredis.FakeStrictRedis"
+
 herd.CACHE_HERD_TIMEOUT = 2
 
 if sys.version_info[0] < 3:
@@ -342,6 +345,8 @@ class DjangoRedisCacheTests(TestCase):
         self.assertFalse(bool(res))
 
     def test_incr(self):
+        if FAKE_REDIS:
+            raise unittest.SkipTest("FakeRedis doesn't support eval")
         try:
             self.cache.set("num", 1)
 
@@ -374,6 +379,8 @@ class DjangoRedisCacheTests(TestCase):
             print(e)
 
     def test_incr_error(self):
+        if FAKE_REDIS:
+            raise unittest.SkipTest("FakeRedis doesn't support eval")
         try:
             with self.assertRaises(ValueError):
                 # key not exists
@@ -396,6 +403,8 @@ class DjangoRedisCacheTests(TestCase):
         self.assertEqual(res, False)
 
     def test_decr(self):
+        if FAKE_REDIS:
+            raise unittest.SkipTest("FakeRedis doesn't support eval")
         try:
             self.cache.set("num", 20)
 
@@ -470,6 +479,9 @@ class DjangoRedisCacheTests(TestCase):
         cache.close()
 
     def test_ttl(self):
+        if FAKE_REDIS:
+            raise unittest.SkipTest("FakeRedis ttl is broken, see https://github.com/jamesls/fakeredis/issues/119")
+
         cache = caches["default"]
         _params = cache._params
         _is_herd = (_params["OPTIONS"]["CLIENT_CLASS"] ==
@@ -505,6 +517,9 @@ class DjangoRedisCacheTests(TestCase):
         self.assertEqual(ttl, 0)
 
     def test_persist(self):
+        if FAKE_REDIS:
+            raise unittest.SkipTest("FakeRedis ttl is broken, see https://github.com/jamesls/fakeredis/issues/119")
+
         self.cache.set("foo", "bar", timeout=20)
         self.cache.persist("foo")
 
@@ -518,6 +533,8 @@ class DjangoRedisCacheTests(TestCase):
         self.assertAlmostEqual(ttl, 20)
 
     def test_lock(self):
+        if FAKE_REDIS:
+            raise unittest.SkipTest("FakeRedis doesn't support locks")
         lock = self.cache.lock("foobar")
         lock.acquire(blocking=True)
 

--- a/tests/runtests-fakeredis.py
+++ b/tests/runtests-fakeredis.py
@@ -1,0 +1,16 @@
+# -*- coding: utf-8 -*-
+
+import os, sys
+sys.path.insert(0, '..')
+os.environ.setdefault("DJANGO_SETTINGS_MODULE", "test_fakeredis")
+
+
+if __name__ == "__main__":
+    from django.core.management import execute_from_command_line
+    args = sys.argv
+    args.insert(1, "test")
+    if len(args) == 2:
+        args.insert(2, "redis_backend_testapp")
+        args.insert(3, "hashring_test")
+
+    execute_from_command_line(args)

--- a/tests/test_fakeredis.py
+++ b/tests/test_fakeredis.py
@@ -1,0 +1,60 @@
+import fakeredis
+
+DATABASES = {
+    "default": {
+        "ENGINE": "django.db.backends.sqlite3"
+    },
+}
+
+SECRET_KEY = "django_tests_secret_key"
+TIME_ZONE = "America/Chicago"
+LANGUAGE_CODE = "en-us"
+ADMIN_MEDIA_PREFIX = "/static/admin/"
+STATICFILES_DIRS = ()
+
+MIDDLEWARE_CLASSES = []
+
+CACHES = {
+    "default": {
+        "BACKEND": "django_redis.cache.RedisCache",
+        "LOCATION": [
+            "redis://127.0.0.1:6379?db=1",
+            "redis://127.0.0.1:6379?db=1",
+        ],
+        "OPTIONS": {
+            "CLIENT_CLASS": "django_redis.client.DefaultClient",
+            "REDIS_CLIENT_CLASS": "fakeredis.FakeStrictRedis",
+        }
+    },
+    "doesnotexist": {
+        "BACKEND": "django_redis.cache.RedisCache",
+        "LOCATION": "127.0.0.1:56379:1",
+        "OPTIONS": {
+            "CLIENT_CLASS": "django_redis.client.DefaultClient",
+            "REDIS_CLIENT_CLASS": "fakeredis.FakeStrictRedis",
+        }
+    },
+    "sample": {
+        "BACKEND": "django_redis.cache.RedisCache",
+        "LOCATION": "127.0.0.1:6379:1,127.0.0.1:6379:1",
+        "OPTIONS": {
+            "CLIENT_CLASS": "django_redis.client.DefaultClient",
+            "REDIS_CLIENT_CLASS": "fakeredis.FakeStrictRedis",
+        }
+    },
+    "with_prefix": {
+        "BACKEND": "django_redis.cache.RedisCache",
+        "LOCATION": "127.0.0.1:6379:1",
+        "OPTIONS": {
+            "CLIENT_CLASS": "django_redis.client.DefaultClient",
+            "REDIS_CLIENT_CLASS": "fakeredis.FakeStrictRedis",
+        },
+        "KEY_PREFIX": "test-prefix",
+    },
+}
+
+INSTALLED_APPS = (
+    "django.contrib.sessions",
+    "redis_backend_testapp",
+    "hashring_test",
+)

